p2p/src/sync/mock_consensus.rs

```diff
@@ -0,0 +1,869 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): A. Altonen
+#![allow(clippy::new_without_default, clippy::result_unit_err)]
+use common::{
+    chain::{
+        block::{consensus_data::ConsensusData, Block, BlockHeader},
+        transaction::Transaction,
+    },
+    primitives::{id, Id, Idable},
+};
+use parity_scale_codec::{Decode, Encode};
+use rand::Rng;
+use std::collections::BTreeMap;
+use tokio::sync::{mpsc, oneshot};
+
+pub type Hash = u64;
+pub type Amount = u64;
+
+trait IdForBlockHeader {
+    fn get_hdr_id(&self) -> Id<Block>;
+}
+
+impl IdForBlockHeader for BlockHeader {
+    fn get_hdr_id(&self) -> Id<Block> {
+        Id::new(&id::hash_encoded(self))
+    }
+}
+
+#[derive(Clone)]
+pub struct Storage<T> {
+    pub store: BTreeMap<Id<Block>, T>,
+}
+
+impl<T> Storage<T> {
+    pub fn new() -> Self {
+        Self {
+            store: BTreeMap::new(),
+        }
+    }
+
+    pub fn add(&mut self, id: Id<Block>, data: T) {
+        self.store.insert(id, data);
+    }
+
+    pub fn remove(&mut self, id: &Id<Block>) -> Option<T> {
+        self.store.remove(id)
+    }
+
+    pub fn get(&self, id: &Id<Block>) -> Option<&T> {
+        self.store.get(id)
+    }
+
+    pub fn get_mut(&mut self, id: &Id<Block>) -> Option<&mut T> {
+        self.store.get_mut(id)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct BlockIndex {
+    pub blkid: Id<Block>,
+    pub prev_blkid: Option<Id<Block>>,
+    pub next_blkid: Option<Id<Block>>,
+    pub trust: u64,
+    pub height: u64,
+}
+
+impl BlockIndex {
+    pub fn new(
+        blkid: Id<Block>,
+        prev_blkid: Option<Id<Block>>,
+        next_blkid: Option<Id<Block>>,
+        trust: u64,
+        height: u64,
+    ) -> Self {
+        Self {
+            blkid,
+            prev_blkid,
+            next_blkid,
+            trust,
+            height,
+        }
+    }
+}
+
+fn get_genesis() -> Block {
+    Block::new(vec![], None, 1337u32, ConsensusData::None).unwrap()
+}
+
+#[derive(Clone)]
+pub struct Consensus {
```

Comment:
It's probably a good idea to use the ConsensusInterface as a trait. I'll be looking into that with Lukas.

Reply:
I'm thinking about this again and unless I'm missing something obvious, I think I'm either way forced to use a wrapper type that forwards the calls to the appropriate consensus implementation or use the subsystem wrapper in syncing tests at which point I might as well just use the proper consensus implementation and remove the mock implementation.

The problem is that the subsystem handle I get should implement an interface that internally calls some implementation, i.e., the issue is that I'm forced to give `SyncManager` a `subsystem::Handle<T>` type, not what `T` actually is. In other words, does it help to have Consensus as an interface if I anyway need to provide abstraction over it?

I'm already in the process of rewriting the syncing tests because there are very infrequent timeouts because of the way they are implemented. I might as well write them from the ground up using the subsystem handler and the actual consensus implementation.

Thoughts? @TheQuantumPhysicist @iljakuklic